<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Martin Lysy, Wookjung P. Kim, Terin Robinson" />

<meta name="date" content="2017-01-09" />

<title>Bayesian Testing of Equal Genotype Proportions between Multiple Populations</title>




<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Bayesian Testing of Equal Genotype Proportions between Multiple Populations</h1>
<h4 class="author"><em>Martin Lysy, Wookjung P. Kim, Terin Robinson</em></h4>
<h4 class="date"><em>2017-01-09</em></h4>



<p>This tutorial shows how to use the <strong>MADPop</strong> package to test for genetic differences between two populations, of which the species contain a variable number of alleles.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">require</span>(MADPop)</code></pre></div>
<div id="pre-processing" class="section level2">
<h2>Pre-Processing</h2>
<p>Our data consists of <span class="math inline">\(N = 215\)</span> recordings of Major Histocompatibility Complex (MHC) genotypes of lake trout from <span class="math inline">\(K = 11\)</span> lakes in Ontario, Canada. For each of the fish, between 1-4 alleles in the MHC genotype are recorded. This is partially because duplicate genes are undetectable by current instrumentation, and possibly because the fish possess a variable number of alleles at the given MHC locations.</p>
<p>Our dataset <code>fish215</code> is included with <strong>MADPop</strong>. A random sample from it looks like this:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">head</span>(fish215[<span class="kw">sample</span>(nObs),])</code></pre></div>
<pre><code>##             Lake  A1  A2  A3  A4
## 75  Michipicoten     r.5     r.4
## 192    Macdonald     z.1        
## 6          Hogan     t.1 r.8 r.1
## 73  Michipicoten             r.7
## 1          Hogan r.3     r.1 r.6
## 108        Slate         v.3 r.4</code></pre>
<p>The first column is the lake name (or population ID) for each sample, the remaining four columns are for potentially recorded allele codes (<code>A1</code>-<code>A4</code>). Here the code to identify a unique allele is a small letter followed by a number, but it could have been the sequence of integers <span class="math inline">\(1, 2,\ldots, A\)</span>, which for the <code>fish215</code> data is <span class="math inline">\(A = 57\)</span> unique alleles.</p>
<p>It is relatively straightforward to import a CSV file into the format above. An example of this is given along with our raw data in the <strong>extdata</strong> directory of the local copy of the <strong>MADPop</strong> package.</p>
</div>
<div id="two-population-comparisons" class="section level2">
<h2>Two-Population Comparisons</h2>
<p>Suppose that we wish to compare two lakes, say Dickey and Simcoe. The allele counts in these lakes are in the table below. It is a subset of the full contingency table on all <span class="math inline">\(K = 11\)</span> lakes, which is produced by the <strong>MADPop</strong> function <code>UM.suff</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">popId &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;Dickey&quot;</span>, <span class="st">&quot;Simcoe&quot;</span>)          <span class="co"># lakes to compare</span>
Xsuff &lt;-<span class="st"> </span><span class="kw">UM.suff</span>(fish215)             <span class="co"># summary statistics for dataset</span>
ctab &lt;-<span class="st"> </span>Xsuff$tab[popId,]             <span class="co"># contingency table</span>
ctab &lt;-<span class="st"> </span>ctab[,<span class="kw">colSums</span>(ctab) &gt;<span class="st"> </span><span class="dv">0</span>] <span class="co"># remove alleles with no counts</span>
<span class="co">#ctab</span>
<span class="kw">rbind</span>(ctab, <span class="dt">Total =</span> <span class="kw">colSums</span>(ctab))</code></pre></div>
<pre><code>##        1.20 1.4.5 1.4.9.45 1.5 10 20 20.54 27.28.29 3 4 4.10 4.11 4.20.27
## Dickey    1     1        0   2  0  2     1        1 0 0    1    0       1
## Simcoe    0     0        1   1  2  0     0        0 1 1    2    1       0
## Total     1     1        1   3  2  2     1        1 1 1    3    1       1
##        4.20.31 4.20.45 4.27 4.33 4.5 4.53.55 4.56 4.7 4.7.10 5 5.20 5.30
## Dickey       1       1    1    0   1       1    1   0      0 0    1    1
## Simcoe       0       0    0    1   1       0    0   3      1 1    0    0
## Total        1       1    1    1   2       1    1   3      1 1    1    1
##        5.32 5.7 7 9.11
## Dickey    1   1 0    0
## Simcoe    0   1 2    1
## Total     1   2 2    1</code></pre>
<p>The unique allele identifiers are encoded as integers between <span class="math inline">\(1\)</span> and <span class="math inline">\(A\)</span> and separated by dots. The original allele names are stored in <code>Xsuff$A</code>, such that the genotype of the first column <code>1.20</code> is</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gtype &lt;-<span class="st"> </span><span class="kw">colnames</span>(ctab)[<span class="dv">1</span>]
gtype &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">strsplit</span>(gtype, <span class="st">&quot;[.]&quot;</span>)[[<span class="dv">1</span>]])
gtype</code></pre></div>
<pre><code>## [1]  1 20</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">names</span>(gtype) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;A&quot;</span>, gtype)
<span class="kw">sapply</span>(gtype, function(ii) Xsuff$A[ii])</code></pre></div>
<pre><code>##    A1   A20 
## &quot;r.1&quot; &quot;u.4&quot;</code></pre>
<p>There are <span class="math inline">\(C = 29\)</span> genotype combinations observed in these two lakes, corresponding to each column of the table.</p>
<div id="multinomial-model" class="section level3">
<h3>Multinomial Model</h3>
<p>In the two-population problem we have <span class="math inline">\(K = 2\)</span> lakes with <span class="math inline">\(N_1\)</span> and <span class="math inline">\(N_2\)</span> fish sampled from each. Let <span class="math inline">\({\boldsymbol Y}_k = (Y_{k1}, \ldots Y_{kC})\)</span> denote the counts for each genotype observed in lake <span class="math inline">\(k\)</span>, such that <span class="math inline">\(\sum_{i=1}^C Y_{ki} = N_k\)</span>. The sampling model for these data is <span class="math display">\[
{\boldsymbol Y}_k {\stackrel{\textrm{ind}}{\sim}}\textrm{Multinomial}(N_k, {\boldsymbol \rho}_k), \quad k = 1,2,
\]</span> where <span class="math inline">\({\boldsymbol \rho}_k = (\rho_{k1},...,\rho_{kC})\)</span> are the population proportions of each genotype, and <span class="math inline">\(\sum_{i=1}^C \rho_{ki} = 1\)</span>.</p>
</div>
<div id="hypothesis-testing" class="section level3">
<h3>Hypothesis Testing</h3>
<p>Our objective is to test <span class="math display">\[
\begin{split}
H_0 &amp;: \textrm{The two populations have the same genotype proportions} \\
&amp; \phantom{: } \iff {\boldsymbol \rho}_1 = {\boldsymbol \rho}_2.
\end{split}
\]</span> The classical test statistics for assessing <span class="math inline">\(H_0\)</span> are Pearson’s Chi-Square statistic <span class="math inline">\(\mathcal X\)</span> and the Likelihood Ratio statistic <span class="math inline">\(\Lambda\)</span>, <span class="math display">\[
\mathcal X = \sum_{k=1}^2 \sum_{i=1}^C \frac{(N_k\hat\rho_i - Y_{ki})^2}{N_k\hat\rho_i}, \qquad \Lambda = 2 \sum_{k=1}^2 \sum_{i=1}^C Y_{ki} \log\left(\frac{Y_{ki}}{N_k\hat\rho_i}\right), \qquad \hat \rho_i = \frac{Y_{1i} + Y_{2i}}{N_1 + N_2}.
\]</span> Under <span class="math inline">\(H_0\)</span>, the asymptotic distribution of either of these test statistics <span class="math inline">\(T = \mathcal X\)</span> or <span class="math inline">\(\Lambda\)</span> is <span class="math inline">\(\chi^2_{(C-1)}\)</span>, such that the <span class="math inline">\(p\)</span>-value <span class="math display">\[
{\textrm{p}_\textrm{v}}= \textrm{Pr}(T &gt; T_\textrm{obs} \mid H_0)
\]</span> for an observed value of <span class="math inline">\(T_\textrm{obs}\)</span> can be estimated as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># observed values of the test statistics</span>
chi2.obs &lt;-<span class="st"> </span><span class="kw">chi2.stat</span>(ctab) <span class="co"># Pearson's chi^2</span>
LRT.obs &lt;-<span class="st"> </span><span class="kw">LRT.stat</span>(ctab) <span class="co"># LR test statistic</span>
T.obs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="dt">chi2 =</span> chi2.obs, <span class="dt">LRT =</span> LRT.obs)
<span class="co"># p-value with asymptotic calculation</span>
C &lt;-<span class="st"> </span><span class="kw">ncol</span>(ctab)
pv.asy &lt;-<span class="st"> </span><span class="kw">pchisq</span>(<span class="dt">q =</span> T.obs, <span class="dt">df =</span> C<span class="dv">-1</span>, <span class="dt">lower.tail =</span> <span class="ot">FALSE</span>)
<span class="kw">signif</span>(pv.asy, <span class="dv">2</span>)</code></pre></div>
<pre><code>##  chi2   LRT 
## 0.330 0.041</code></pre>
<p>The Chi-Square and LR tests are asymptotically equivalent and so should give roughly the same <span class="math inline">\(p\)</span>-values. The huge discrepancy observed here indicates that the sample sizes are too small for asymptotics to kick in. A more reliable <span class="math inline">\(p\)</span>-value estimate can be obtained by the Bootstrap method, which in this case consists of simulating <span class="math inline">\(M\)</span> contigency tables with <span class="math inline">\({\boldsymbol Y}_k^{(m)} {\stackrel{\textrm{ind}}{\sim}}\textrm{Multinomial}(N_k, \hat{{\boldsymbol \rho}})\)</span>, <span class="math inline">\(m = 1, \ldots, M\)</span>, where <span class="math inline">\(\hat{{\boldsymbol \rho}}\)</span> is the estimate of the common probability vector <span class="math inline">\({\boldsymbol \rho}_1 = {\boldsymbol \rho}_2\)</span> under <span class="math inline">\(H_0\)</span>. For each contingency table <span class="math inline">\(({\boldsymbol Y}_1^{(m)}, {\boldsymbol Y}_2^{(m)})\)</span>, we calculate the test statistic <span class="math inline">\(T^{(m)}\)</span>, and the bootstrapped p-value is defined as <span class="math display">\[
{\textrm{p}_\textrm{v}^{\textrm{boot}}}= \frac 1 M \sum*{m=1}^M \delta(T^{(m)} \ge T_\textrm{obs}).
\]</span> <span class="math inline">\({\textrm{p}_\textrm{v}^{\textrm{boot}}}\)</span> is calculated with <strong>MADPop</strong> as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">N1 &lt;-<span class="st"> </span><span class="kw">sum</span>(ctab[<span class="dv">1</span>,])                     <span class="co"># size of first sample</span>
N2 &lt;-<span class="st"> </span><span class="kw">sum</span>(ctab[<span class="dv">2</span>,])                     <span class="co"># size of second sample</span>
rho.hat &lt;-<span class="st"> </span><span class="kw">colSums</span>(ctab)/(N1+N2)        <span class="co"># common probability vector</span>
<span class="co"># bootstrap distribution of the test statistics</span>
<span class="co"># set verbose = TRUE for progress output</span>
<span class="kw">system.time</span>({
  T.boot &lt;-<span class="st"> </span><span class="kw">UM.eqtest</span>(<span class="dt">N1 =</span> N1, <span class="dt">N2 =</span> N2, <span class="dt">p0 =</span> rho.hat, <span class="dt">nreps =</span><span class="fl">1e4</span>,
                      <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
})</code></pre></div>
<pre><code>##    user  system elapsed 
##    1.01    0.00    1.01</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># bootstrap p-value</span>
pv.boot &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">t</span>(T.boot) &gt;=<span class="st"> </span>T.obs)
<span class="kw">signif</span>(pv.boot, <span class="dv">2</span>)</code></pre></div>
<pre><code>##   chi2    LRT 
## 0.0074 0.0069</code></pre>
<p>Note that the bootstrap <span class="math inline">\(p\)</span>-values for both tests are roughly the same and decisively reject <span class="math inline">\(H_0\)</span>, whereas the less reliable asymptotic <span class="math inline">\(p\)</span>-values both failed to reject (at quite different significance levels).</p>
</div>
</div>
<div id="pairwise-comparisons-between-multiple-populations" class="section level2">
<h2>Pairwise Comparisons between Multiple Populations</h2>
<p>Bootstrapping overcomes many deficiencies of the asymptotic <span class="math inline">\(p\)</span>-value calculation. However, bootstrapping has a tendency to reject <span class="math inline">\(H_0\)</span> when sample sizes are small. To see why this is, consider all columns of <code>ctab</code> which have only one genotype count between the two lakes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">itab1 &lt;-<span class="st"> </span><span class="kw">colSums</span>(ctab) ==<span class="st"> </span><span class="dv">1</span>             <span class="co"># single count genotypes</span>
<span class="kw">cbind</span>(ctab[,itab1],
      <span class="dt">Other =</span> <span class="kw">rowSums</span>(ctab[,!itab1]),
      <span class="dt">Total =</span> <span class="kw">rowSums</span>(ctab))</code></pre></div>
<pre><code>##        1.20 1.4.5 1.4.9.45 20.54 27.28.29 3 4 4.11 4.20.27 4.20.31 4.20.45
## Dickey    1     1        0     1        1 0 0    0       1       1       1
## Simcoe    0     0        1     0        0 1 1    1       0       0       0
##        4.27 4.33 4.53.55 4.56 4.7.10 5 5.20 5.30 5.32 9.11 Other Total
## Dickey    1    0       1    1      0 0    1    1    1    0     7    20
## Simcoe    0    1       0    0      1 1    0    0    0    1    12    20</code></pre>
<p>There are <span class="math inline">\(c_1 = 21\)</span> such columns, accounting for <span class="math inline">\(\hat p_1 = 0.525\)</span> of the common genotype distribution under <span class="math inline">\(H_0\)</span>, as estimated from the two-lake sample. For each of these columns, observing counts in one lake but not the other provides evidence against <span class="math inline">\(H_0\)</span>. Moreover, under the estimated common distribution <span class="math inline">\(\hat {{\boldsymbol \rho}}\)</span>, it is very unlikely to have counts in only one of the lakes for each of these <span class="math inline">\(c_1 = 21\)</span> genotypes. Therefore, the data appear to provide very strong evidence against <span class="math inline">\(H_0\)</span>. However, it is not so unlikely to have <span class="math inline">\(c_1 = 21\)</span> one-count genotypes if the true number of unique genotypes in these two lakes is much larger than the observed value of <span class="math inline">\(C = 29\)</span>. With <span class="math inline">\(C = 29\)</span> unique genotypes in only <span class="math inline">\(N = N_1 + N_2 = 40\)</span> fish samples, it is quite plausible that a new sample of fish would yield several genotypes which are not present in the original two-lake sample <code>ctab</code>.</p>
<p>One way to obtain information about the unobserved genotypes in lakes Dickey and Simcoe is to consider the genotypes in <em>all</em> <span class="math inline">\(K = 11\)</span> Ontario lakes for which we have collected data. A natural way to do this is through a hierarchical model: <span class="math display">\[
\begin{split}
{\boldsymbol Y}_k \mid {\boldsymbol \rho}_k &amp; {\stackrel{\textrm{ind}}{\sim}}\textrm{Multinomial}(N_k, {\boldsymbol \rho}_k), \quad k = 1,\ldots,K \\
{\boldsymbol \rho}_k &amp; {\stackrel{\textrm{iid}}{\sim}}\textrm{Dirichlet}({\boldsymbol \alpha}), \qquad {\boldsymbol \alpha}= (\alpha_1, \ldots, \alpha_C).
\end{split}
\]</span> That is, each population is allowed to have its own probability vector <span class="math inline">\({\boldsymbol \rho}_k\)</span>. However, these probability vectors are drawn from a common Dirichlet distribution. The common distribution specifies that before any data is drawn, we have <span class="math display">\[
E[{\boldsymbol \rho}] = \bar {\boldsymbol \alpha}, \qquad {\textrm{var}}(\rho_i) = \frac{(\bar \alpha_i)(1-\bar \alpha_i)}{\alpha_0 + 1},
\]</span> where <span class="math inline">\(\alpha_0 = \sum_{i=1}^C \alpha_i\)</span> and <span class="math inline">\(\bar{{\boldsymbol \alpha}} = {\boldsymbol \alpha}/ \alpha_0\)</span>. Moreover, the posterior distribution of <span class="math inline">\({\boldsymbol \rho}_k\)</span> given <span class="math inline">\({\boldsymbol \alpha}\)</span> and the data is also Dirichlet: <span class="math display">\[
{\boldsymbol \rho}_k \mid {\boldsymbol Y}{\stackrel{\textrm{ind}}{\sim}}\textrm{Dirichlet}({\boldsymbol \alpha}+ {\boldsymbol Y}_k).
\]</span> In this sense, the posterior estimate of the probability vector <span class="math inline">\({\boldsymbol \rho}_k\)</span> can be non-zero even for genotypes for which <span class="math inline">\(Y_{ki} = 0\)</span>, as long as <span class="math inline">\(\alpha_i &gt; 0\)</span>. In practice, we estimate <span class="math inline">\({\boldsymbol \alpha}\)</span> from the data <span class="math inline">\({\boldsymbol Y}\)</span> from all <span class="math inline">\(K\)</span> populations (details in the following section). The extent to which the genotypes observed in one lake affect inference in the other lakes is determined by <span class="math inline">\(\alpha_0\)</span> (the larger this value, the larger the effect).</p>
<div id="parameter-estimation" class="section level3">
<h3>Parameter Estimation</h3>
<p>The likelihood function for this model corresponds to a Dirichlet-Multinomial distribution, <span class="math display">\[
\begin{split}
\mathcal L({\boldsymbol \alpha}\mid {\boldsymbol Y}) = \prod_{k=1}^K p({\boldsymbol Y}_k \mid {\boldsymbol \alpha}) &amp; = \prod_{k=1}^K \int p({\boldsymbol Y}_k \mid {\boldsymbol \rho}_k) p({\boldsymbol \rho}_k \mid {\boldsymbol \alpha}) \,\mathrm{d}{\boldsymbol \rho}_k \\
&amp; = \prod_{k=1}^K \left[\frac{N_k!\cdot\Gamma(\alpha_0)}{\Gamma(N_k+\alpha_0)} \prod_{i=1}^C \frac{\Gamma(Y_{ki} + \alpha_i)}{Y_{ki}!\cdot\Gamma(\alpha_i)}\right],
\end{split}
\]</span> from which <span class="math inline">\({\boldsymbol \alpha}\)</span> can be estimated by maximum likelihood <span class="citation">(Minka 2000)</span>. Alternatively we estimate <span class="math inline">\({\boldsymbol \alpha}\)</span> using a Bayesian approach, which is more readily extensible to more complex genotype models, for which <span class="math inline">\(\mathcal L({\boldsymbol \alpha}\mid {\boldsymbol Y})\)</span> has no closed form (see Future Work).</p>
<p>First, we specify a prior distribution <span class="math display">\[
\pi(\alpha_0, \bar {\boldsymbol \alpha}) = \frac{1}{(1+ \alpha_0)^2},
\]</span> which is a uniform distribution on the prior probability vector <span class="math inline">\(\bar {\boldsymbol \alpha}\)</span>, and a uniform distribution on the variance-like parameter <span class="math inline">\(A = (1+\alpha_0)^{-1}\)</span>, in the sense that the prior mean and variance of a given genotype population probability are <span class="math display">\[
E[\rho_{kj}] = \tilde \alpha_j, \qquad {\textrm{var}}(\rho_{kj}) = A \cdot \tilde \alpha_j(1-\tilde \alpha_j).
\]</span> Then, we sample from the posterior distribution <span class="math inline">\(p({\boldsymbol \alpha}\mid {\boldsymbol Y}) \propto {\mathcal L}({\boldsymbol \alpha}\mid {\boldsymbol Y}) \pi({\boldsymbol \alpha})\)</span> using a Markov chain Monte Carlo (MCMC) algorithm provided by the <strong>stan</strong> programming language <span class="citation">(Stan Development Team 2016)</span>, as detailed below.</p>
</div>
</div>
<div id="bayesian-hypothesis-testing" class="section level2">
<h2>Bayesian Hypothesis Testing</h2>
<p>Under the hierarchical model, the null hypothesis of equal genetic proportions between two populations, say <span class="math inline">\(k = 1,2\)</span>, is <span class="math inline">\(H_0: {\boldsymbol \rho}_1 = {\boldsymbol \rho}_2 = {\boldsymbol \rho}_{12}\)</span>. Our Bayesian hypothesis-testing strategy is implemented in two steps:</p>
<ol style="list-style-type: decimal">
<li>Sample from the posterior distribution <span class="math inline">\(p({\boldsymbol \alpha}\mid {\boldsymbol Y}, H_0)\)</span>.</li>
<li>Sample from the posterior distribution of either classical test statistic, <span class="math inline">\(T = \mathcal X\)</span> or <span class="math inline">\(\Lambda\)</span>, thus obtaining a <em>posterior p-value</em> <span class="math display">\[
{\textrm{p}_\textrm{v}^{\textrm{post}}}= \textrm{Pr}(T &gt; T_\textrm{obs} \mid {\boldsymbol Y}, H_0).
\]</span></li>
</ol>
<div id="mcmc-sampling-from-the-posterior-distribution" class="section level3">
<h3>MCMC Sampling from the Posterior Distribution</h3>
<p>In order to estimate <span class="math inline">\({\boldsymbol \alpha}\)</span> under <span class="math inline">\(H_0\)</span>, we apply the Dirichlet-Multinomial distribution to <span class="math inline">\(K-1\)</span> lakes, with the first two being collapsed into a common count vector <span class="math inline">\({\boldsymbol Y}_{12} = {\boldsymbol Y}_1 + {\boldsymbol Y}_2\)</span> with sample size <span class="math inline">\(N_{12} = N_1 + N_2\)</span>. MCMC sampling is implemented via a Hybrid Monte Carlo (HMC) variant provided by the <strong>R</strong> package <strong>rstan</strong>. In <strong>MADPop</strong>, sampling from <span class="math inline">\(p({\boldsymbol \alpha}\mid {\boldsymbol Y}, H_0)\)</span> is accomplished with the function <code>hUM.post</code>, where “hUM” stands for <em>hierarchical Unconstrained Multinomial</em> model. We start by merging the two samples from equal distributions under <span class="math inline">\(H_0\)</span>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">popId                                   <span class="co"># equal lakes under H0</span></code></pre></div>
<pre><code>## [1] &quot;Dickey&quot; &quot;Simcoe&quot;</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">eqId0 &lt;-<span class="st"> </span><span class="kw">paste0</span>(popId, <span class="dt">collapse =</span> <span class="st">&quot;.&quot;</span>)  <span class="co"># merged lake name</span>
popId0 &lt;-<span class="st"> </span><span class="kw">as.character</span>(fish215$Lake)    <span class="co"># all lake names</span>
popId0[popId0 %in%<span class="st"> </span>popId] &lt;-<span class="st"> </span>eqId0
<span class="kw">table</span>(popId0, <span class="dt">dnn =</span> <span class="ot">NULL</span>)               <span class="co"># merged lake counts</span></code></pre></div>
<pre><code>##       Crystal Dickey.Simcoe         Hogan     Kingscote     Macdonald 
##            20            40            21            18            19 
##       Manitou  Michipicoten       Opeongo        Seneca         Slate 
##            20            20            20            18            19</code></pre>
<p>Next, we sample from <span class="math inline">\(p({\boldsymbol \alpha}\mid {\boldsymbol Y}, H_0)\)</span> using <strong>rstan</strong>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">X0 &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="dt">Id =</span> popId0, fish215[-<span class="dv">1</span>])  <span class="co"># allele data with merged lakes</span>

nsamples &lt;-<span class="st"> </span><span class="fl">1e4</span>
fit0 &lt;-<span class="st"> </span><span class="kw">hUM.post</span>(<span class="dt">nsamples =</span> nsamples, <span class="dt">X =</span> X0,
                 <span class="dt">rhoId =</span> eqId0,     <span class="co"># output rho only for merged lakes</span>
                 <span class="dt">chains =</span> <span class="dv">1</span>,  <span class="co"># next two arguments are passed to rstan</span>
                 <span class="dt">warmup =</span> <span class="kw">min</span>(<span class="fl">1e4</span>, <span class="kw">floor</span>(nsamples/<span class="dv">10</span>)),
                 <span class="dt">full.stan.out =</span> <span class="ot">FALSE</span>)</code></pre></div>
<pre><code>## 
## SAMPLING FOR MODEL 'DirichletMultinomial' NOW (CHAIN 1).
## 
## Chain 1, Iteration:    1 / 10000 [  0%]  (Warmup)
## Chain 1, Iteration: 1000 / 10000 [ 10%]  (Warmup)
## Chain 1, Iteration: 1001 / 10000 [ 10%]  (Sampling)
## Chain 1, Iteration: 2000 / 10000 [ 20%]  (Sampling)
## Chain 1, Iteration: 3000 / 10000 [ 30%]  (Sampling)
## Chain 1, Iteration: 4000 / 10000 [ 40%]  (Sampling)
## Chain 1, Iteration: 5000 / 10000 [ 50%]  (Sampling)
## Chain 1, Iteration: 6000 / 10000 [ 60%]  (Sampling)
## Chain 1, Iteration: 7000 / 10000 [ 70%]  (Sampling)
## Chain 1, Iteration: 8000 / 10000 [ 80%]  (Sampling)
## Chain 1, Iteration: 9000 / 10000 [ 90%]  (Sampling)
## Chain 1, Iteration: 10000 / 10000 [100%]  (Sampling)
##  Elapsed Time: 4.377 seconds (Warm-up)
##                42.908 seconds (Sampling)
##                47.285 seconds (Total)</code></pre>
<p><code>rstan</code> typically throws a number of exceptions which are usually harmless. The <code>MADPop</code> package lists <code>rstan</code> as a dependency, such that its sophisticated tuning mechanism is exposed. Optionally, <code>hUM.post</code> returns the entire <code>rstan</code> output (<code>full.stan.output = TRUE</code>), which can be run through <code>rstan</code>’s MCMC convergence diagnostics.</p>
<p>By setting the <code>rhoId</code> argument of <code>hUM.post</code>, the <code>fit0</code> object contains samples from <span class="math inline">\(p({\boldsymbol \alpha}, {\boldsymbol \rho}_{12} \mid {\boldsymbol Y}, H_0)\)</span>. Boxplots of the 50 highest posterior probability genotypes in the common distribution <span class="math inline">\({\boldsymbol \rho}_{12}\)</span> are displayed below:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">rho.post &lt;-<span class="st"> </span>fit0$rho[,<span class="dv">1</span>,]   <span class="co"># p(rho12 | Y)</span>
<span class="co"># sort genotype counts by decreasing order</span>
rho.count &lt;-<span class="st"> </span><span class="kw">colSums</span>(Xsuff$tab[popId,])
rho.ord &lt;-<span class="st"> </span><span class="kw">order</span>(<span class="kw">colMeans</span>(rho.post), <span class="dt">decreasing =</span> <span class="ot">TRUE</span>)

<span class="co"># plot</span>
nbxp &lt;-<span class="st"> </span><span class="dv">50</span>            <span class="co"># number of genotypes for which to make boxplots</span>
clrs &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;#999999&quot;</span>, <span class="st">&quot;#E69F00&quot;</span>, <span class="st">&quot;#56B4E9&quot;</span>, <span class="st">&quot;#009E73&quot;</span>, <span class="st">&quot;#F0E442&quot;</span>, <span class="st">&quot;#0072B2&quot;</span>,
          <span class="st">&quot;#D55E00&quot;</span>, <span class="st">&quot;#CC79A7&quot;</span>)
rho.ct &lt;-<span class="st"> </span>rho.count[rho.ord[<span class="dv">1</span>:nbxp]]
rho.lgd &lt;-<span class="st"> </span><span class="kw">unique</span>(<span class="kw">sort</span>(rho.ct))
rho.box &lt;-<span class="st"> </span>rho.post[,rho.ord[<span class="dv">1</span>:nbxp]]
<span class="kw">par</span>(<span class="dt">mar =</span> <span class="kw">c</span>(<span class="dv">5</span>,<span class="fl">4.5</span>,.<span class="dv">5</span>,.<span class="dv">5</span>)+.<span class="dv">1</span>, <span class="dt">cex.axis =</span> .<span class="dv">7</span>)
<span class="kw">boxplot</span>(<span class="dt">x =</span> rho.box,
        <span class="dt">las =</span> <span class="dv">2</span>, <span class="dt">col =</span> clrs[rho.ct<span class="dv">+1</span>], <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">cex =</span> .<span class="dv">2</span>)
<span class="kw">title</span>(<span class="dt">xlab =</span> <span class="st">&quot;Genotype&quot;</span>, <span class="dt">line =</span> <span class="dv">4</span>)
<span class="kw">title</span>(<span class="dt">ylab =</span> <span class="kw">expression</span>(<span class="kw">p</span>(<span class="kw">bold</span>(rho)[(<span class="dv">12</span>)]*<span class="st">&quot; | &quot;</span>*<span class="kw">bold</span>(Y))))
<span class="kw">legend</span>(<span class="st">&quot;topright&quot;</span>, <span class="dt">legend =</span> <span class="kw">rev</span>(rho.lgd), <span class="dt">fill =</span> <span class="kw">rev</span>(clrs[rho.lgd<span class="dv">+1</span>]),
       <span class="dt">title =</span> <span class="st">&quot;Counts&quot;</span>, <span class="dt">ncol =</span> <span class="dv">2</span>, <span class="dt">cex =</span> .<span class="dv">8</span>)</code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAqAAAAFQCAMAAABXmDxzAAAAsVBMVEUAAAAAADoAAGYAOjoAOpAAZrYAnnM6AAA6ADo6AGY6Ojo6OpA6ZmY6ZrY6kJA6kNtWtOlmAABmADpmAGZmOjpmOpBmZrZmkJBmkNtmtrZmtv+QOgCQOjqQOmaQZgCQZpCQkDqQkLaQtpCQ27aQ29uQ2/+ZmZm2ZgC2Zjq2Zma2kJC2/7a2///bkDrbkGbbkJDbtrbb25Db///mnwD/tmb/tpD/25D/29v//7b//9v///90llzaAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAZ1klEQVR4nO2dC5vbunGGuY6dteumPutN2lTyadOkitKeVS8bVZH4/39YeceAHBAgLuSQ/L7nnDVEYAYg+Ao3QmSWQ5BgZUsXAILGBEAh0QKgkGgBUEi0ACgkWgAUEi0ACokWAIVEC4BCogVAIdFKBmgGQb6aBdBUjiFRSnGdASgUTQAUEi0ACokWAIVEC4BComW4zrfnLPsUw+f8gGZAd0vir+b1w1v+OBoIvX15c/cJQKEgsVfz/nIo/t4+n1gT4YBCmxJ7ncsGtFbR1T+dKiSL/29f/pBl3+4v5aFrVv61+wSgUJB4QD++14GyKS1obQF9/tZ+un8/qVRjPgEoFKRxQBsWW0A/n+qG9K0ZBDj4BKBQkMa7+JLUx/HQdfHdPwWhWXZw8AlAoSCNTJKKv70WlACa05HqiE8ACgWJv86XYgL0OH58b8ag5T8XNRQte/rifwAKpZfhOheT9Gqhvp7F55cs+4fXFtDHsejcMYuHZhFudUKiBUAh0QKgkGgBUEi0ACgkWgAUEi0AConW/gDFftFVCYBCorU/QKFVCYBCokWuM/vomuFNd/4RN/SXIAAUiiYK6K8b0Ws/3LaU/bWVxsj95QMAhaLLBmip8vcdJBkL6CX7e7SgUHy5AHrRfnzEA/o/7+jioQSyA3p77o1B+S4eY1AohVxaUKcxKACFUijWGBSAQkkEQCHRsq2Dlt377as2ScI6KDSfrNf5kmWmX2/qAqBQAuFWJyRaABQSLQAKiRYAhURr04Bib/L6Zd1uN7SwpgOgUDRRQH9qpF3W8nUKB83iz61ouscxw3Y7KL5sgFbPt9N2i/CAnj+V/w18LgAoms1NyQZo8whbasEBWt4N7VbqASgUTdYuPs97z/tmAW2fW9/3KbCLB8CrkgugvQ3LABSaTw6AXvR78WsHFFqV7IBeek9SXiWgaDbXKus66KW/l4ldB404Sbp2y1r318Lh2bCbato6KABdq2wXzvQ+xL6iLTM1LxXJmx8yP34Y8sdC/T5ku4Tnqqk8WP3cXyIt1N++vtdQPn7+5fWt/2JQ+20sPfXU3H094auQSuLuxV+Ldvh8qIJlF3/91TP/XnCXdVAAun5JBzT/23t+OVgy0SPAypYkDtCui28ALXT5ZskE2q7EAaomSXUXXzSoEVpQr2Y1yWABmiZxgNbLTFXj2S4zcQ0oAN2J5AHqkYkeAZa2pP0BinZxVQKgkGhlKUT9Jyu4+Yy4IAR1kg0o1tx3LzG7mdjNIjZqAejmJRtQS0po+1ohoNCeJHAMOgFQsLx5AdC5tcYyLyjZgG7xYrqf08Yrwk1ixqAkdeDa02YEQPOlAfXq4vd7sfYo2YCCxehaW5WuG9C11bYAra3KBALKx1vs11bvkKMEAurVxWNcsFFtBVDe6Wa0xXNykxHQ8mG4Wfbk9iiICZnoEWgsR7SbEx2TAdAaz1iIxm1BQ28/rUZbPKfJ4gG9dI8eeRz538F5Z6JHAFBoXCygF+35IOxP3b0z0SPcWcM66T7FAXr7vZ7mv0J7+Rgt6BZZ9Bp370wMoI9/fdfTDA6EZKJHLDGLl3OxAahdbBd/fwkfd5oy0SOi3UkSeftpv1jFkwHQSOtLTCZ6RGAXHzhJSg0QAA0XP4u/Zhn/IMXwTPQInwk7G7/ywQBQNsm0UH+JtUjfd6xH+LSg882sAOjyMt/qrJ/XzD90PiATPWIJQAHDmmQE9HFcClBbvEhAQX0imQA9r6KLt3jaZB+9s68CD+hlyUlSKKC2nFau7ZyJk1a4zBQ434dWpdUs1HuNQaHVyzyLT5SJHpECUGhLWg2ggUbWIgJwmVohoKx96Cw+dJIFwBNpNYAmvpMUOslazT0ridmPad2AujudoCWu1sIN9NoAvb9oj7EPvpnkDSib1GsI4CU57ZpggqZp8onIbkG9jOJdVzmARky6rDYGqPtw0AvQJCOARVhZDaCTtRpA402SAhdPvQAVSfUqtGtAvRQKqFevMCH/jWk1gHoZJQY0cWO5cUDdTmQ1gC7Rgi7SMW9mNBDnRMYBrXfVB2+8SwaoxZO1XLM1hhO0GgBtinMiY4BeskP179WIaPW++ErV++LVR3MmekS0FtTm1D371M2uuzaDaojMgD6Oas/d9SP76Ib799PtS7WKf3/58EY+jmSiRywLqLVQ0Q56Kd6CbmolLF7QGPT29f3xo9zZ/Pj5l6IF7T427thXKuu5zzYGDe2t52tW3b+JNv/zUS0V0GvR9Z8PVbDs4slHcyZ6xBKAes2sAhUKqJd/4c2um0YBvT2Xo89qeMlKPKCBOdmE4WZ6jQH6OB7y64c3M6CkT78Pu3izYxWxLKATFK/jDUy5N40BWpF5f/3FCCiZFZVpl5wkeWGzSMcLTdJoC1o1h/ffstBVqtaVGo6XXWaaD1CvlLy5V0kW1twltYxBD8Xfa7L9oGSGrwVJAiYYOpy0YRF4CdYDaOBYZx4FzeJ9MjHGiJmQezWWgUZeS2OhivdNTLgGshdAQydRgTOv0HOSqHnWQADo1Hiv6xJxiCwcW3eFA9r+NCndb5IW6eK9CAjEJuJ6gRhA5ynIaAt6/x7nCU3TzmS+FnSCka2kFoWyJoZKIgGAmjaJhGTikFo2oBGHk16j3UDF+9LNo4XHoGzqhQG1FCrJfCdJA8/63w6gj38j4bA3JQkYg8abOaXOyZJ9KKBr00gLeu3e0nkJfFroJEAz25p9EmxomRIYsSlDz2kXGu3iL812ztCnhU5rQfNA1twvphfVNlfzfRV2oYXHoLMBmgRANt5mlLokGxMAjWPEKnC+E9FIotyKNzoGjfUmhYkVlbiLT0z1IjlxeQazmhjwUEAv5dAzfCfT0PEUQzEtqKVMixSPVcSBq+30Z5ER0HI7faH7P0bOZJrhsr31woMJW0nY+NRd/NxtsRHQ++/qpc//TvmmOX4Mqg6RX4SuBlDWfkK7loTq+RS7JOYWtFmb/4/5AeUPLtvFxzMiSjECmSA5LeyYP0M4z/+vWpy//SZyJsYYADo1J1YTAFmkhZ0qM6Dt22Rn3263cUDjGbFapFl1N5rs3gxoRAFQ4YC6uw8dY07IqUk1DNNdIrX+M3DT3cQ6SzxeCzRiy7QdQCMmtZh7A5pfD3qaS+/zZMVomsUAKtLIJi8jzjyi/AHNz9rukHPwm2UB6HyA8pd9CarjuGcBzS9Zu5X+/hK8l2kM0MySSAwBXh2vxC5+EUC9vlRNHB++PbdT+Bgvjp8IKH9wLYAubGRTqJGXpwDUDYC2iMbAMwqg6qbShNtLcrCRAyhRYGMbD1CPFjSuYrSglu14crAhWqKLnyBbuxY4mHDPCYAu3q4tYkTk1RjG+ypY5Ado9diGJX52bEjOANrr97P2qEopEBs5DXzORc8HqJt7I6DXavgZ+nM5xnG4E+dmdT3YhBrl4yn5eJuRO6BeKIcB2mwHzc8x2tB038RlAaUFkQ2ozalFtpTu34rJKJsAbZ96E2VLvblQbGPoZNmPZ/r9xbGRCCif1KLUX++xrA3htgW9pG1BLb21D6B8Y7oe1hITMIFar+KxniwaS2kCNL9WT/O+8M/0nqgogLLxFqcSsUlsRDShhbXET2j33Lt4N/cmQNtHL4rZDwpAHeNzj4NeRl6DCVvKoakhHFUAdJVGrAJYm5ayNTCEo2qiYz65BWDWSDgBEo1YhbLoNVhoDAzhqIriGIDOYkRkiZ8wwnD3NIwzhKMqmWMAGseIlwUmL0DdM2rjDOGoMjtmG0Mny34836xKISC1Uc6lZI+GtqApjNYGqI1a1n6HgBIJHxfYjEa0HkDZeIvT1bAmp4t3763jUb2NFtQDUK8tTnKw2Y3RbgHVWM2HwTVezC0aCQZ0QnILwBYjidcFRsODAy0IqLqXqoJOTgDogkb5eEq/nEYUBmj3gvg6cDbctzfA8+tGWRecXAIAOosRqxUAev9+un156wKPH4bt9/6A2opkaVbJMdkEbNwo51K6KQjQ29f3Gso6UMNKyjLecxsBtbDGFx+ACjbKPVK2cYawk66fim790AWuv3rm37oQBVCWRd4xd2w1F3OHRnMBmv/tnX/Q2HKAtkE87UGyUTJAe118eejCPclpAqC9YUEsQLWQlOsCo+HBgYIA7U2SynY0tAUlB41OJlDLxOPH9OKMkgFary7dX9/UMhP7KDxRgFqozb2oFU6AcKN0gDoqCqAWz2GA8kGJF3OLRisG1FYkG2teRqIv5haNVgpohPuffkZtcNIjIFzHBTanNk/CWdsVoD/VyvpO3AH0KyOTEzuz0r49DNX8dExPafPkRr0c1gCoQznHWfMsuGuzm8QodOUBgI4LgCYxksgaAM3tF5PN2YtaADqLkVRAuf2giwHKB92NLJ4mFM9WJtGs7RNQLmXW85wa0IUFQPOEFyYY0J/YoAHVaMtM0iWRtU0B6jwGNQDKtbAu5+EO6IQacU/qXjxbQoGsAVAW0L82clvdt5UxcFyQxIiTRNa2Dyjp7b0A9V2bshV8YUC5oETWACgAFc0aAI0AqDu1AHQWo1UBGjoG9QWUD7obWTwZqsDDEwCNL09AB23pkoAuLACabxJQS3HWAygniawB0FkBnVAj7kltKZ09SWRtV4CGjUF7q/cTyhg4LkhixEkiawDUHVB10Pvk5ADKBSWytilAh3fgJ20WAaACWdsSoCTGtlmEB1SFpgHq3IcC0HmM5ANKDngBamlWezm5Y8OW1D0l1kEnHxxoYUCHSSeMQd2bVQvVvTImq4cAMcVTZRfE2vYBZRtD2yTJBugwaNgC1aNWuASytmFAh7291yzeC1DuoHPBJ5yjU0ofT2JY2xqgY+O1eMtMXoCSttQ2mOAORjSy1x5JB0A9tE5AVVu6tJGJ2vGBKaUbgI7KE1B1XXYOqAr9udFwPGSpUgA6Jj9AVTD0VufeAc20dlVFA1CrYzYGgEZvQWUvngoG1M2Sm1AAUNevtx4EoJZMIlgOx6gsoMMrvAtAxytywswJgHokMphySzbcNNgwI+YPbhJQrS0dH5gCUD2GHUWFZLorQN3rjDHSUG2DmQ1gAOrqxOA6FNAuJMjIAVAbtROM7ACzB23xI0ajxZUIqHv84GDwGDQxa15G3rP4Warc22i9LSgANRkZK9JWe15GANRwMARQcl29AHUfGc5nxFLtUDu2awtAwy0DamuTgKrO3tYsh10Cvyr3PjiSs3BA2YN7A1SNQSmgg4NafG9ckOwqp5EcQBNWnPsgkBkXrB/QCGtTy0nOGNSpwZ9RpN2xAGyYWQ0PLg4oe07VyQ7OmRxdVOsBVEJtUbHUsvE2I+tXQYVCAWWNhv4nrDxM6pQs8Ww1G8JR5Q/oElNKwUbsFZ6PauZgVCNOUgClX6G1YbOI0QRAh6zuBtD6NfFdoPs4kol2nGvYteZ+bdjMZ+QHaJgRd1B0F3//frp9eesC3cexTKbJvYWVgc18Rl7LTHsD9Pb1/fHj1AW6j407S9Z2sbbkoApmvSAbL8zIyZPRiNQte7EtB/cB6PVTnp8PXaD7OJYJBE0RAIVEK2EXH+AYghoJnyRBe1eEZab765vnMhME2RQGqEcmEDRFABQSLQAKidY8gEKQr+YAdARdLijjXiKM5jNyJA+AwgiAGrMUU1swmt0IgMJItBEAhZFoIwAKI9FGABRGoo0AKIxEGwFQGIk2kgsoBLkLgEKiBUAh0QKgkGgBUEi0ACgkWgAUEi0ACokWAIVEC4BCojU7oNfqRydPp/z+8q09Vv7K/pJlH9+1YJeSJCVGlZqnmiinxJNKqjwRaTk9/fE5+/CmJb19PhUfyoM0U5X09px9qr30/NdlIu65syfxfMrHsUzZZdw4Zc+JlFmViS3e7fnpTy/1iSqnXJloPDHqgv3r0Ehlyl086om9JJzmBvRxPJT/3L6+31//0NZUcUblM56KgySoUuYqKTV6qX9gVXymTomnNinxRERy+nG6PR+qeJW0fqxPz1OukpYprx/eyotBsldlIiXhzp7Esynzv5zy22/yyyftRNlzImXWyzQo3uPn8kkw+f1378QpWyYar4xUkFRJ7/TqTNmLRz2xl4TT3IBWX+nqn/K/l/qpJEXw8q2OU0GVMldJiVHxffxGWqvOKfGkjDpPGkEqp6LuWiPlqQCk5ylXSauU99dfNCO9TF1Jukz1c1LZ8ykLR01jSJwy50TLrJWpX7wmie6ULZMe3xqpIK0SrUqbTI0X77VfzzZglmpBvzTncK47jI//+/u8+Wa3QZUyV0mJUaHL079X1404JZ66pMRT+USppiSDnOg3u0j6l1N5fVpIm0xV0oaY3355o0aqTKQkKlN6Tl08n7JwUnwVmiamc8qdkwrSMnHFa5ow3SlbJhqvjFSQXgdVpSpT9uJRT7TKRjX7GLQYpqjhIjn6dCoGfVqwS0mS9r5yj+PTSXNKzElSEp1f1VhP5VT+uTbmbdJ6ENiOizubLmnZ1edlj6b778pEz6nLlKZU8VzKsmE6NAc7p/w5qaBWJqZ4xQiajEEbp3yZVDwx6oLadVBVqjLlLh71pFXZmDCLh0RrcUDJF6wXJPPQESOzlP2Ip/7RbpZOZ7TajLidxZOpv+ZsOA3mp/5dPD+LH5S5HbqRpI17OncmyxFa8ehDXG317FLlpvUCmqn24FimF+QfLNvT4oAaReahgfYTPKlZOpnR0hmxmsWrqb8SOw3mp/7afIif7w/LrJL21jCaEyXLESpIko7Xk1fl8rVjydNWJqLZZ/HOJSPzUC+nvSllM6O1VRydpZP1ADIjVvHd1J+yNpwG81N/bZpsmO8Pzp4k7a1htCdKliPaIEnqXs+WijKuF7SZjudpKxPR7C2omvOV4stYHSXzUGLNdoc9p9R1NaUknnop+/mrWTqZ0Q5nxPrUXzllp8H81F/FG2fxg7PX/Xfu+3Nnshyhl4TWXtcb93NqqmT8Oo2uF+jF69l3IS3erPm7+HbOx39HyVEyD1XR7KK2NjfvRKeUypM2TR7mr03o1Sy7syezeJVSm8cOp8H81J/Gs/P94dn3/Cvz7kRJ6Unx2qR02KKGAMpcq5Lx62RYL6CZtisPqlfQPXXLHaNacAxKOilScmt/1PWh8fKXLfdRkdWTNmzphgBKbJWE1pOy9/G0EKB1xXStvNafjLX9eh/qnpMeaoPGnMaMHOLjGdUDPGYA41hmPZ4OW46H6mBXkWNV4lBPo+ek7HueBN5J0tW18loXPdr2a31otPxlixvA+IkMW7QhgBJbJaH1pOwne5K7zARB+UKAzjfyY1fCDQvI7afhSnhvb1jfqW0THL+3TO2nI8vb6mBvZ5zTiQ7PrpcTq0FODh0vPXs2ODjltSzUxxvxu+fEL5/zRqaV8Lxvf7ds7OODRN1+Orrkrw6qrWkTTpQ/O7Vzj9OEnEie3AZJErRtp5O7UD/j3JlbCbdlb14J79vfbRv7uGBva1q9n05fXlcHianjifJnp5yaTtk5J66e2CCtMkuZLVqgi3dcoY2XE798zopfCT8eyribaXm9vgzGTXDahjZta1q9n44ur6uD2s44xxPlz07fuTcQzWls04Kpnmx7JU3bF8Qu1M85d2ZWwm3Z8yvhdEA1cEouKLsJjgbVhFztpyMTanVQ3xnneqLs2ek79wZSOU25Lc9tkGT3Shq3L4hfqF9Y842GV6MJ2x8mOR1uX3DXfgF1XP8u5dbz3fWta3yQMeL3Fzj3tnZxGwNZ8RsARmS8JUCKz29fcNaOAXVe/3bv+cy/WhuxZ/cXhG42pK64jYG8DBsAPPI8quLz2xdctWdAXeXe8xl3oY3Zs/sLIva2Xj1reJ6jxRe8zLRCOfV8/S2CfJAzYvcXTO5tR0p2PJT/TOtZnRxP3CtJ5Ty+AqAOsvR8/BZBPsgZsfsLovW2fj2rRV57JTW5jq8AaAR53XxYy24/XnzpE5wTAHXW2CBu4oY0zsg0yZ9azLnEn3L0uzAANI68bj6sZLefQXzpY58TAIVEC4A6ymtoRYwM9m0f3o82PD5ufwKgdnndFKUblyw/4eV/NM09Pm6HAqAOCp2lW+y5aLWFb+cCoE7ympwSI4v9MFpt4Zua68YEQN0UOku32A+iBa8vzSsAupCw289NAHQpWe5GR9xut2oB0MU0ejc64na7dQuAylSSze1rFACVqYjb7dYtACpUEbfbrVoAFBItAAqJFgCFRAuAQqIFQCHRAqCQaAFQSLQAKCRaABQSLQAKiRYAhUQLgEKiBUAh0QKgkGgBUEi0ACgkWgAUEi0ACokWAIVEC4BCogVA06h6cIjlMezXwzxlWbUAaBJdygfTPY6jv8m8vxxmKs2aBUBTqP7NcEGo4RXYlQCoiwBoCp2bzv1yKP8UnX35oq6Xfy5f6pqXryYsD5SvW/34LyXBl0+3z398rhJ1qaFGADSBHkeC2OXpVD2i9v5SBMoP14LS+8unqgUtH8zwOB4KWA/V8S411AiAJhDtvOuHzZfPASsDRd9f03t9OpWpyv9vn081kpeP713qZQouUAA0gWpAz9XTP6/NG+MP1cHiTz0+bQ+cP5VcNs+5eTp1qRcquTwB0ATqXtFRNIXX+jm1WQ/QIlAdKFKcv9UvK6wAbVMvVXRxAqAp1E6Srm0LmjfN6qAFvX//0/cTAXTNb/ZKIgCaQm0fXQ0962AHqDYGLWZIf1fAXI9Bz+UY9LBEeQULgCbRpVoqOpdP8S7n5fm54bGauXez+G9tyvJ1x2oWX6ZetviCBEDTqLrVWU/GL+3rvA5NM3ptljrP5b3Qqne/Pf/TczPwvODFCpoA6MK6fX3HtH1EAHRhXcq2FIAaBUAX1e25mu8DUKMAKCRaABQSrf8H+BMx1Xvn9p0AAAAASUVORK5CYII=" /><!-- --></p>
</div>
<div id="calculating-the-posterior-p-value" class="section level3">
<h3>Calculating the Posterior P-Value</h3>
<p>This is calculated analogously to the bootstrapped p-value, by generating <span class="math inline">\(M\)</span> contingency tables with <span class="math inline">\({\boldsymbol Y}_k^{(m)} {\stackrel{\textrm{ind}}{\sim}}\textrm{Multinomial}(N_k, {\boldsymbol \rho}_{12}^{(m)}\)</span>, <span class="math inline">\(m = 1, \ldots, M\)</span>. However, for each contingency table, the common probability vector <span class="math inline">\({\boldsymbol \rho}_{12}^{(m)}\)</span> is different, corresponding to a random draw from <span class="math inline">\(p({\boldsymbol \rho}_{12} \mid {\boldsymbol Y}, H_0)\)</span>. The test statistic <span class="math inline">\(T^{(m)}\)</span> is then calculated and we have <span class="math display">\[
{\textrm{p}_\textrm{v}^{\textrm{post}}}= \frac 1 M \sum_{m=1}^M \delta(T^{(m)} \ge T_{\textrm{obs}}).
\]</span> <span class="math inline">\({\textrm{p}_\textrm{v}^{\textrm{post}}}\)</span> is calculated with <strong>MADPop</strong> as follows:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">system.time</span>({
  T.post &lt;-<span class="st"> </span><span class="kw">UM.eqtest</span>(<span class="dt">N1 =</span> N1, <span class="dt">N2 =</span> N2, <span class="dt">p0 =</span> rho.post, <span class="dt">nreps =</span> <span class="fl">1e4</span>,
                      <span class="dt">verbose =</span> <span class="ot">FALSE</span>)
})</code></pre></div>
<pre><code>##    user  system elapsed 
##    1.45    0.00    1.47</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># posterior p-value</span>
pv.post &lt;-<span class="st"> </span><span class="kw">rowMeans</span>(<span class="kw">t</span>(T.post) &gt;=<span class="st"> </span>T.obs)</code></pre></div>
<p>We can now compare the three types of p-values (asymptotic, bootstrapped, posterior) for each test statistic (<span class="math inline">\(\mathcal X\)</span> and <span class="math inline">\(\Lambda\)</span>):</p>
<table>
<caption>p-value (<span class="math inline">\(\times100\%\)</span>)</caption>
<thead>
<tr class="header">
<th align="left"></th>
<th align="right">Asymptotic</th>
<th align="right">Bootstrap</th>
<th align="right">Posterior</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><span class="math inline">\(\mathcal X\)</span></td>
<td align="right">33.0</td>
<td align="right">0.74</td>
<td align="right">14</td>
</tr>
<tr class="even">
<td align="left"><span class="math inline">\(\Lambda\)</span></td>
<td align="right">4.1</td>
<td align="right">0.69</td>
<td align="right">13</td>
</tr>
</tbody>
</table>
<p>We see that <span class="math inline">\({\textrm{p}_\textrm{v}^{\textrm{post}}}\)</span> is much more conservative than <span class="math inline">\({\textrm{p}_\textrm{v}^{\textrm{boot}}}\)</span>.</p>
</div>
</div>
<div id="future-work" class="section level2">
<h2>Future Work</h2>
<p>Here we used a completely unconstrained Multinomial model for the genotype counts, in the sense that the only restriction on <span class="math inline">\({\boldsymbol \rho}_k\)</span> is that <span class="math inline">\(\rho_{ki} \ge 0\)</span> and <span class="math inline">\(\sum_{i=1}^C \rho_{ki} = 1\)</span>. However, it is possible to impose genetic constraints such as Hardy-Weinberg equilibrium, preferential mating, etc., which effectively reduce the degrees of freedom of <span class="math inline">\({\boldsymbol \rho}_k\)</span> below <span class="math inline">\(C-1\)</span>. In this case, the closed-form likelihood <span class="math inline">\({\mathcal L}({\boldsymbol \alpha}\mid {\boldsymbol Y})\)</span> is typically unavailable, but the <strong>stan</strong> code can easily be modified to sample from <span class="math inline">\(p({\boldsymbol \alpha}, {\boldsymbol{\mathcal R}}\mid {\boldsymbol Y})\)</span>, where <span class="math inline">\({\boldsymbol{\mathcal R}}= ({\boldsymbol \rho}_1, \ldots, {\boldsymbol \rho}_K)\)</span>. We hope to feature some of these extensions to the basic Dirichlet-Multinomial model in the next version of <strong>MADPop</strong>.</p>
<!-- ```{r, echo = FALSE} -->
<!-- # kable(as.data.frame(as.list(count0)))   # counts in each lake -->
<!-- ``` -->
</div>
<div id="references" class="section level2 unnumbered">
<h2>References</h2>
<div id="refs" class="references">
<div id="ref-minka00">
<p>Minka, T.P. 2000. “Estimating a Dirichlet Distribution.” <em>Technical Report</em>. Massachusetts Institute of Technology.</p>
</div>
<div id="ref-rstan">
<p>Stan Development Team. 2016. “Rstan: The R Interface to Stan.” <a href="http://mc-stan.org" class="uri">http://mc-stan.org</a>.</p>
</div>
</div>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
